### DISCUSSION TITLE
Most efficient way to store float tensors in duckdb?

### DISCUSSION THREAD

**Discussion by webshared** [duckdb/duckdb#6211] [General]
I need to store training dataset for my model, which means a multiple float vectors per row. 
The goal is to pour random training samples into GPUs as fast as possible.  Read queries will be mostly `SELECT * FROM train USING SAMPLE ...` 

I tried a few options: 
* `list(NUMBER)` is extremely slow, I discarded the idea. Looks like it serializes and deserializes each element which is not optimal. 

* Using BLOB fields and save binary data works but depends drastically on the blob size. Here are my measurements: 

150k rows in database 

8088 bytes BLOB column
481Mb DB filesize
**64 secs to save
6 secs to load 256 rows batch**

1348 bytes BLOB column
433Mb DB filesize
**3.7 secs to save
0.2 secs to load 256 rows batch**

Thus, increasing a single column BLOB length from 1348 to 8088 slows down write and read performance 20x times! 
Any ideas on how to fix that?

**Comment by Mytherin** [comment]
I would actually recommend using lists for this. DuckDB stores lists internally as columns - which should in general be much faster to store, operate on and compress than blobs. It is possible that the reservoir sample operator is not yet optimized for use with lists - you could try other sampling operators, e.g. [bernoulli or system sampling](https://duckdb.org/docs/sql/samples.html):

```sql
SELECT * FROM train USING SAMPLE bernoulli(0.1%);
SELECT * FROM train USING SAMPLE system(0.1%);
```

As for the blobs - DuckDB uses a different storage mechanism for large blobs (>4KB) as they might not fit into the pages the database uses internally. This is similar to [TOAST](https://www.gojek.io/blog/a-toast-from-postgresql) if you are familiar with that from Postgres. We currently use gzip to (de)compress large blobs which adds extra cost as well. We haven't really done much work in optimizing this path as it is a rather uncommon use case - so I would recommend sticking to storing data in smaller blobs or storing data as lists for now.

**Reply by webshared** [reply]
`bernoulli(0.1%)` takes the same 6 secs per batch, but system is not stable - `system(0.1%)` returns empty rowset, same for 1%. system(2%) returned 2048 rows. 

I am going to try storing train sets in a simple small table without large blobs, and store blobs in another table. Then `select * from blobs where id IN (...)` using the sampled batch blob ids.

**Reply by Mytherin** [reply]
I meant to suggest using other sampling methods in combination with lists. Lists should be the fastest solution here. If they are not it would be interesting for us to know in which operation they are slow exactly.

**Reply by webshared** [reply]
I tried using list with C++ API: 

```
  void appendTensorAsList(torch::Tensor t)
  {
    auto listValues = vector<duckdb::Value>(t.numel());
    const float *ptr = (const float *)t.data_ptr();
    for (auto i = 0; i < t.numel(); ++i)
      listValues[i] = duckdb::Value::FLOAT(ptr[i]);
    appender.Append(duckdb::Value::LIST(listValues));
  }

  torch::Tensor tensor_value;
  tensor_value = ... 

  appender.BeginRow();
  ...
  appendTensorAsList(tensor_value);
  appender.EndRow();
```

It took 42 secs to save 150k rows with 8KB of tensor data each. Better than 64s when using 8KB blob, but way worse that I'd expect assuming it takes 7.5 secs for 4KB blob.  
Read speed improved a lot - now it takes 0.6 secs per 256 items batch. 

Maybe there is more effective way to write `torch::tensor` or `vector<float>` in C++?

**Reply by Mytherin** [reply]
Appending using the `Value` API is indeed very slow, so that is not too surprising. We don't really have any convenience methods for appending lists efficiently from within C++, but you could append them using our DataChunks. This is a bit more involved as it involves directly constructing DuckDB's internal structures.

Here is an example of how you might do that:

```cpp

TEST_CASE("Test appending lists", "[api]") {
	DuckDB db(nullptr);
	Connection con(db);
	// create the table
	REQUIRE_NO_FAIL(con.Query("CREATE TABLE float_lists(i FLOAT[])"));
	// now create the appender
	Appender appender(con, "float_lists");

	DataChunk chunk;
	vector<LogicalType> insert_types;
	insert_types = { LogicalType::LIST(LogicalType::FLOAT) };
	chunk.Initialize(Allocator::DefaultAllocator(), insert_types);

	// appending 2 lists, one with 1500 elements, one with 5000 elements
	vector<float> my_tensor1;
	for(idx_t i = 0; i < 1500; i++) {
		my_tensor1.push_back(i);
	}
	vector<float> my_tensor2;
	for(idx_t i = 0; i < 5000; i++) {
		my_tensor2.push_back(i * 3);
	}

	auto &list_vector = chunk.data[0];
	auto list_data = FlatVector::GetData<list_entry_t>(list_vector);
	ListVector::Reserve(list_vector, my_tensor1.size() + my_tensor2.size());
	ListVector::SetListSize(list_vector, my_tensor1.size() + my_tensor2.size());
	auto &child_vector = ListVector::GetEntry(chunk.data[0]);
	auto child_data = FlatVector::GetData<float>(child_vector);

	list_data[0].offset = 0;
	list_data[0].length = my_tensor1.size();
	list_data[1].offset = list_data[0].offset + list_data[0].length;
	list_data[1].length = my_tensor2.size();
	memcpy(child_data, my_tensor1.data(), sizeof(float) * my_tensor1.size());
	memcpy(child_data + list_data[1].offset, my_tensor2.data(), sizeof(float) * my_tensor2.size());

	// 2 lists
	chunk.SetCardinality(2);

	// appending works initially
	appender.AppendDataChunk(chunk);
	chunk.Reset();

	appender.Flush();

	// now create a new table with the same name but different types
	con.Query("SELECT * FROM float_lists")->Print();
}
``` 

```
[ Rows: 2]
[0.0, 1.0, 2.0, 3.0, ... 1497.0, 1498.0, 1499.0]
[0.0, 3.0, 6.0, 9.0, ... 14991.0, 14994.0, 14997.0]
```

**Reply by webshared** [reply]
> Appending using the `Value` API is indeed very slow, so that is not too surprising. We don't really have any convenience methods for appending lists efficiently from within C++, but you could append them using our DataChunks. This is a bit more involved as it involves directly constructing DuckDB's internal structures.

Looks good, but how stable is this API? Is it going to change anytime soon?

**Comment by l1t1** [comment]
try test more sizes

**Reply by arjenpdevries** [reply]
I have not looked into this in detail but you might also be interested in mixing duckdb and lance:
https://blog.eto.ai/peeking-duck-duckdb-lance-for-computer-vision-100224d8a999

**Reply by webshared** [reply]
I've got interesting result - db size is bigger while I'm saving 4044 bytes per blob vs. 8088 bytes, and still fast enough. 

4044 bytes BLOB column
657Mb DB filesize
7.5 secs to save
0,56 secs to load 256 rows batch

Likely >4KB blob triggers gzip compression which makes it take less space but also slows it down a lot as @Mytherin suggested.

**Comment by l1t1** [comment]
it seems that 1348 bytes is the best size
